pkg/kv/kvserver/client_raft_test.go

```diff
@@ -6324,3 +6325,531 @@ func TestRaftCheckQuorum(t *testing.T) {
+		IncomingRaftMessageHandler: secStore,
+		OutgoingRaftMessageHandler: secStore,
+	}
+
+	tc.Servers[0].RaftTransport().Listen(firstStore.StoreID(), filterRaftHandler1)
+	tc.Servers[0].RaftTransport().ListenOutgoingMessage(firstStore.StoreID(), filterRaftHandler1)
+	tc.Servers[1].RaftTransport().Listen(secStore.StoreID(), filterRaftHandler2)
+	tc.Servers[1].RaftTransport().ListenOutgoingMessage(secStore.StoreID(), filterRaftHandler2)
+
+	key := tc.ScratchRange(t)
+	require.NoError(t, tc.WaitForSplitAndInitialization(key))
+	tc.AddVotersOrFatal(t, key, tc.Target(1))
+	require.NoError(t, tc.WaitForVoters(key))
+
+	firstStore, fromReplica := getFirstStoreReplica(t, tc.Server(0), key)
+	secStore, toReplica := getFirstStoreReplica(t, tc.Server(1), key)
+
+	fromReplicaDesc := roachpb.ReplicaDescriptor{
+		ReplicaID: fromReplica.ReplicaID(),
+		NodeID:    tc.Server(0).NodeID(),
+		StoreID:   firstStore.StoreID(),
+	}
+	toReplicaDesc := roachpb.ReplicaDescriptor{
+		ReplicaID: toReplica.ReplicaID(),
+		NodeID:    tc.Server(1).NodeID(),
+		StoreID:   secStore.StoreID(),
+	}
+
+	request := &kvserverpb.RaftMessageRequest{
+		FromReplica: fromReplicaDesc,
+		ToReplica:   toReplicaDesc,
+		Heartbeats: []kvserverpb.RaftHeartbeat{
+			{
+				RangeID:       1,
+				FromReplicaID: fromReplicaDesc.ReplicaID,
+				ToReplicaID:   toReplicaDesc.ReplicaID,
+			},
+		},
+	}
+
+	metricsNames := []string{
+		"raft.rcvd.bytes",
+		"raft.rcvd.cross_region.bytes",
+		"raft.rcvd.cross_zone.bytes",
+		"raft.sent.bytes",
+		"raft.sent.cross_region.bytes",
+		"raft.sent.cross_zone.bytes"}
+	storeMetrics0Before := getFirstStoreMetrics(t, tc, 0, metricsNames)
+	storeMetrics1Before := getFirstStoreMetrics(t, tc, 1, metricsNames)
+
+	// Although messages should have been sent as part of the setup, we are
+	// sending additional one to ensure that the test success is not coincidental.
+	if sent := tc.Servers[0].RaftTransport().SendAsync(request, rpc.DefaultClass); !sent {
+		t.Fatalf("unable to send message from server 0 to server 1")
+	}
+	time.Sleep(5 * time.Second)
+	storeMetrics0After := getFirstStoreMetrics(t, tc, 0, metricsNames)
+	storeMetrics1After := getFirstStoreMetrics(t, tc, 1, metricsNames)
+
+	zeroMetrics := map[string]int64{
+		"raft.rcvd.bytes":              0,
+		"raft.rcvd.cross_region.bytes": 0,
+		"raft.rcvd.cross_zone.bytes":   0,
+		"raft.sent.bytes":              0,
+		"raft.sent.cross_region.bytes": 0,
+		"raft.sent.cross_zone.bytes":   0,
+	}
+	// Since all incoming and outgoing messages are dropped, stores should always
+	// have zero metrics.
+	require.Equal(t, zeroMetrics, storeMetrics0Before)
+	require.Equal(t, zeroMetrics, storeMetrics1Before)
+	require.Equal(t, zeroMetrics, storeMetrics0After)
+	require.Equal(t, zeroMetrics, storeMetrics1After)
+}
+
+// TestRaftCrossRegionMetricsNew verifies that raft messages receiving and
+// sending correctly updates the cross-region byte count metrics.
+func TestRaftCrossRegionMetricsNew(t *testing.T) {
```

Comment:
I haven't reviewed the code yet, but this test looks to be testing something basic from a great distance. Is it not possible to have whatever method does the metric update take just the essentials and then write a unit test that verifies it directly? Starting a TestServer is pretty hefty for what this test is doing.

Reply:
Thanks a lot for the valuable feedback! I agree that using unit tests on the metrics helper function seems to make more sense and could get rid of the interceptor usages. Making the changes accordingly to both this and the other PRs.

Reply:
Do you mind giving these two test cases another look? I changed test cases for DistSender and Node metrics to unit tests since they both require interceptors which don’t seem to be worth the confusion. But this test case doesn’t cause any disruption to the main logic and works quite nice. 

The unit tests are essentially pretty trivial and just testing the switch statement. There are still some interesting things that were tested in this end to end test. For example, it tested other logic on the way, such as storepool GetNodeLocality, and also tested our worries about batch response matching the nature of batch requests. I agree that this seems to be a large test considering we are only testing metrics increment. But I thought leaving one end to end test case would be nice.

Reply:
Honestly, I still think these tests are overkill - these tests will be annoying for anyone to engage with should they ever be in the way of anything, and I don't think they'll add benefit over a much simpler unit test either. I know this is a judgment call, but since my team is going to own these, let's just go with some dead simple unit test that makes sure things are hooked up: set up a minimal raft transport with the handler, send a few messages, observe the impact on metrics.